[SystemZ] Add a SystemZ specific pre-RA scheduling strategy.

JonPsson1 opened the pull request:
As the heavy spilling in Cactus was found to be caused by the
MachineScheduler, it has been clear that there is room for improvement in
GenericScheduler. A first attempt was made with #90181, which however did not
materialize in anything actually useful.

This time, a SystemZ specific MachineSchedStrategy has been developed, guided
by experimentation and benchmarking.

Finding that GenericScheduler fails to handle liveness despite all the work
it is doing with PressureSets etc, I tried a simpler approach by keeping
track of live registers and scheduling e.g. a load immediately if it closed a
live range. If this is always done first, it then seems to work well to also
consider latencies as a second step. The final piece that makes it all work
well is the distinction of tiny regions, which remedies seen problems with
comparison elimination, phys-reg copys and even compile time. These regions
(with up to 10 instructions) are mainly left as they were.

SPEC Results (omitting <1% changes):

```
f538.imagick_r      -13.34%
f507.cactuBSSN_r    -11.60% (*)
f526.blender_r       -3.82%
f544.nab_r           -1.99%
f508.namd_r          -1.82%
f519.lbm_r           -1.67%

(*) Improved 5% by disabling pre-RA scheduling.

geometric mean across all benchmarks: -1.81%

Number of Spill/Reload instructions : -5718 (-1%)
Number of register moves:             +1778 (+0.2%)
```

Compile time showed at first some heavy regressions that were due to a
large amount of LIS->getInterval() lookups in functions with many small
blocks (regions) and vregs. With the tiny regions handling however, this
problem went away and it seems to be now on par with GenericSched.
If needed there are likely things that could be sped up.

JonPsson committed:
Rebase + squash. Was 2e34820

JonPsson1 commented:
@preames @wangpc-pp @michaelmaitland @arsenm @asb 

Any comments welcome.

JonPsson1 commented:
Note: 
The test vec-cmpsel.ll has been split into vec-cmpsel-01.ll and vec-cmpsel-02.ll.
vec-move-23.ll has been reformatted a bit by update_llc_test_checks.py

wangpc-pp reviewed (commented):
Sorry I am not an expert on SystemZ (know nothing about it), so I can only give some high level comments.

JonPsson1 commented on llvm/lib/Target/SystemZ/SystemZTargetMachine.cpp:
```diff
@@ -35,6 +35,11 @@ static cl::opt<bool> EnableMachineCombinerPass(
     cl::init(true), cl::Hidden);
 
+static cl::opt<bool> GenericSched(
+    "generic-sched", cl::Hidden, cl::init(false),
```
It actually is: -generic-sched would run GenericScheduler instead of this...

wangpc-pp replied:
Yes, that is what I understand. But here we disable this option, that means we aggressively use the new scheduler without further experiment.

JonPsson1 replied:
Ah, ok. Well, the benchmarking has been done (see above) so at this point it is looking very good, with a few possible improvements ("todos") left for the future...

wangpc-pp replied:
Ah I get it, then it is OK for me.

mshockwave reviewed (commented):
Do you have any insights on why and how the existing pressure set approach is insufficient?

JonPsson1 commented:
> Do you have any insights on why and how the existing pressure set approach is insufficient?

That's a good question... Maybe it is doing too little too late - it waits for certain limits to be reached but then it has already opened up many register intervals and for some reason it isn't enough to try to reduce register pressure at that point. It seems to work better to reduce liveness regardless of the current pressure, that is as soon as possible.

I am not sure if the PSets themselves are overly complex and get in the way of things:  On SystemZ, the simple approach I am using is to consider the fact of basically two sets of registers available: FP regs (which overlap the vector regs), and the GPR regs. The idea is to reduce FP pressure as much as possible by ignoring GPR regs in FP intensive regions. I see 5 PSets on SystemZ, but I am not sure that works the same as having just two, but ideally it should. Prioritizing FP regs seems to work well, and is maybe an important difference to the GenericScheduler.

Another difference is that it seems to work better not to consider memory operands (GPR registers used to hold addresses), as they are not naturally part of the data flow. This is something PSets don't know about - how the register is actually used. But I don't think that is a major point in all this.

I have also seen the handling of critical PSets go wrong with a large region which is very FP intensive. In the *input order*, it is the integer sets that have the most pressure, but there is a lot of potential FP overlap which results after scheduling (with increased heavy spilling). I tried fixing this however by providing the correct critical PSets, but it did not help much on that test case. Another option I haven't tried (suggested by @atrick, I think) is to have a second pass which could among other things maybe correct this if it goes wrong. This I haven't tried however.

JonPsson1 commented:
A patch with scripts that I have been working with as a tool for developing this strategy can be found here: https://github.com/llvm/llvm-project/pull/136483.

@atrick Any comments from you (and others ofc) would be very welcome before committing this. Are there any "no-no":s in this strategy - for instance related to compile time? This strategy is now at a good point, but that doesn't mean that things can get
  better. Do you have any ideas at this point that migh be worth trying?

@lhames

arsenm commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.cpp:
```diff
@@ -2144,6 +2144,34 @@ unsigned SystemZInstrInfo::getFusedCompare(unsigned Opcode,
+    Compare.getOperand(1).isImm() && Compare.getOperand(1).getImm() == 0;
+}
+
+unsigned SystemZInstrInfo::
+getCompareSourceReg(const MachineInstr &Compare) const {
+  unsigned reg = 0;
```
Use Register instead of unsigned

arsenm commented on llvm/lib/Target/SystemZ/SystemZInstrInfo.h:
```diff
@@ -356,6 +356,17 @@ class SystemZInstrInfo : public SystemZGenInstrInfo {
+  // Return true if Compare is a comparison against zero.
+  bool isCompareZero(const MachineInstr &Compare) const;
+
+  // Return the source register of Compare, which is the unknown value
+  // being tested.
+  unsigned getCompareSourceReg(const MachineInstr &Compare) const;
```
Can you use analyzeCompare instead?

JonPsson1 replied:
No, that doesn't consider the Load And Test when used as a comparison. I could however simplify it a bit as this is actually supposed to only be called after isCompareZero() returns true.

arsenm commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp:
```diff
@@ -5,22 +5,421 @@
+      dbgs() << ", ";
+    else
+      First = false;
+    dbgs() << "%" << R.virtRegIndex();
```
printReg

arsenm commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp:
```diff
@@ -5,22 +5,421 @@
+  }
+}
+
+void SystemZPreRASchedStrategy::VRegSet::dump(std::string Msg) {
+  dbgs() << Msg.c_str();
```
Don't need .c_str, and don't need std::string (but really should get rid of the prefix argument, the use context can do that just as well and it's much more usable from the debugger with no string argument)

arsenm commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp:
```diff
@@ -5,22 +5,421 @@
+      if (IsStore)
+        StoresGroup.insert(SU);
+    }
+    else if (IsStore && !StoresGroup.empty() && SU->getDepth() == CurrMaxDepth) {
```
Brace placement

arsenm commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp:
```diff
@@ -5,22 +5,421 @@
+
+  // Find the registers that are live at the bottom, before scheduling.
+  LiveRegs.clear();
+  for (unsigned I = 0, E = DAG->MRI.getNumVirtRegs(); I != E; ++I) {
```
This looks heavy, can you just look at operations in the region?

JonPsson1 replied:
I tried doing it now first "on the fly" - that is the first time count() is called for a register. That however seemed a bit complex to think about, spreading this logic out in various places. So I instead changed this to still doing it in initialize(), but however only for the registers defed/used in the region. I think this is a good guard against some type of cases that could get slow, and it is also a slight speedup compared to before: zig.bc is now slightly faster even (1.65% vs 1.7%), SPEC: 0.95% ave vs 0.91%, same worst case, test-suite: Ave 0.93% vs 0.87%, worst 26.7% vs 26%.

arsenm commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.h:
```diff
@@ -24,6 +34,69 @@
+  bool ShouldReduceLatency;
+
+  // Keep track of currently live registers.
+  struct VRegSet : std::set<Register> {
```
Don't subclass base data structures

JonPsson1 replied:
ok changed from is-a to has-a, a bit more methods to implement but at least useful for adding assertions.

arsenm commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.h:
```diff
@@ -24,6 +34,69 @@
+  // Num instructions left to schedule.
+  unsigned NumLeft;
+
+  // Tru if latency scheduling is enabled.
```
Typo

JonPsson1 commented:
@arsenm Thanks for review - patch updated accordingly.

JonPsson1 commented:
ping - any thoughts / objections before committing this?

JonPsson1 commented:
Rebase.

uweigand commented:
Hi @JonPsson1, I've started looking into this and have a few high-level questions:
- As to latency - common code uses logic in `shouldReduceLatency` based on critical path length to decide whether or not latency is important right now.  What is the difference between that common code method and the custom check you're implementing here?
- As to register pressure - you're now using the pressure *difference* per SU computed by common code, but not the actual *pressure* itself at this point (which is also already computed by common code).  Wouldn't that be important to consider?  It doesn't really matter if an instruction would use one more vector register if we still have 20 unused at this point, does it?

JonPsson1 commented:
> As to latency - common code uses logic in shouldReduceLatency based on critical path length to decide whether or not latency is important right now. What is the difference between that common code method and the custom check you're implementing here?

**shouldReduceLatency()** can give a decision before each picked SU based on the remaining critical path in the given moment. This is based on the current cycle and remaining latency. The **HasDataSequences** I implemented looks at a region before scheduling and then triggers latency reduction across the entire region - before each picked node.

I have generally tried to move away from the cycle-based scheduling heuristics used by GenericScheduler as they mostly make exact sense for an in-order target during pre-RA scheduling, I'd say. I could however not really make the argument for one or the other except by trying it out, which I did:

First, I examined how many times in SystemZPreRASchedStrategy::tryCandidate() at the point of interest (below liveness reduction) these would yield "true":

```
Total                 : 4.8M queries
shouldReduceLatency() : 3M   
HasDataSequence       : 200K 
HasDataSequence && shouldReduceLatency: 42K  
```

In other words, 4.8M times (note that this number includes all the iterations across the Available queue) would be "always true", barred the initial check to exclude some very "wide" DAGs. This is what was done before improving this recently. shouldReduceLatency() cuts that almost in half, while HasDataSequence is relatively rare.

When I noticed a regression after returning to this project on lbm, I tried disabling this latency heuristic alltogether, and saw these results:

```
patch <> patch without latency heuristic
Improvements:
0.965: f519.lbm_r 
0.976: i505.mcf_r 
0.981: f544.nab_r 
Regressions:
1.052: f508.namd_r 
1.051: f507.cactuBSSN_r 
1.026: f526.blender_r 
```

The lbm problem went away, but other benchmarks suffered.

After working with this and adding the (HasDataSequences || Rem.IsAcyclicLatencyLimited) guard (per the patch posted here currently), I see these numbers:

```
main <> patch
Improvements:
0.885: f507.cactuBSSN_r 
0.963: f544.nab_r 
0.974: f526.blender_r 
0.987: f508.namd_r 
0.990: i531.deepsjeng_r 
0.998: i505.mcf_r 
Regressions:
1.007: i523.xalancbmk_r 
1.004: f519.lbm_r

```
Now, back to the experiment with shouldReduceLatency(). I tried two changes: using shouldReduceLatency() instead of HasDataSequences, or both combined like (HasDataSequences && shouldReduceLatency()).

There is a slight (+2000 spills/reloads) increase in spilling with shouldReduceLatency(), which is sort of as expected.

Performancewise, it seems to e.g. handle cactus equally well, but there are slight losses on two benchmarks:

```
patch <> patch using shouldReduceLatecy()
Regressions:
1.019: f526.blender_r 
1.010: i500.perlbench_r 
1.008: i505.mcf_r 
```

```
patch <> patch using (HasDataSequences && shouldReduceLatency())
Regressions:
1.028: f526.blender_r 
1.019: i500.perlbench_r 
```

Given the results, it seems better to schedule more agressively for latency throughout the region than to do it per the cycle based critical path heuristic - if done only on the regions where this matters. I get some confidence in HasDataSequence as it gives better performance with a lot less involvement, and also better than the combination of the two.


```
diff --git a/llvm/include/llvm/CodeGen/MachineScheduler.h b/llvm/include/llvm/CodeGen/MachineScheduler.h
index a0ef475..c21bb33 100644
--- a/llvm/include/llvm/CodeGen/MachineScheduler.h
+++ b/llvm/include/llvm/CodeGen/MachineScheduler.h
@@ -1224,7 +1224,7 @@ protected:
   void traceCandidate(const SchedCandidate &Cand);
 #endif
 
-private:
+protected:
   bool shouldReduceLatency(const CandPolicy &Policy, SchedBoundary &CurrZone,
                            bool ComputeRemLatency, unsigned &RemLatency) const;
 };
diff --git a/llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp b/llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp
index a3df7fc..dff2700 100644
--- a/llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp
+++ b/llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp
@@ -344,7 +344,14 @@ bool SystemZPreRASchedStrategy::tryCandidate(SchedCandidate &Cand,
     // Don't extend the scheduled latency in regions with many nodes in
     // simple data sequences, or for (single block loop) regions that are
     // acyclically (within a single loop iteration) latency limited.
-    if ((HasDataSequences || Rem.IsAcyclicLatencyLimited) &&
+
+    // EXPERIMENT: Use GenericSchedulerBase::shouldReduceLatency() instead of HasDataSequences
+    CandPolicy P;
+    getRemLat(Zone);
+    bool GenericShouldReduceLat = shouldReduceLatency(P, *Zone, false, RemLat);
+    if (((HasDataSequences && GenericShouldReduceLat) || Rem.IsAcyclicLatencyLimited) &&
+ // if ((GenericShouldReduceLat || Rem.IsAcyclicLatencyLimited) &&
+ // if ((HasDataSequences || Rem.IsAcyclicLatencyLimited) &&
         TryCand.SU->getHeight() != Cand.SU->getHeight() &&
         (std::max(TryCand.SU->getHeight(), Cand.SU->getHeight()) >
          Zone->getScheduledLatency())) {
```

JonPsson1 commented:
> As to register pressure - you're now using the pressure difference per SU computed by common code, but not the actual pressure itself at this point (which is also already computed by common code). Wouldn't that be important to consider? It doesn't really matter if an instruction would use one more vector register if we still have 20 unused at this point, does it?

First of all, per the comment in RegisterPressure.cpp: "The register tracker is unaware of global liveness so ignores normal live-thru ranges...", there is (currently) no truly accurate awareness of the number of used / free registers. 

The OOO reasoning I am using to motivate e.g. pulling a VL down right before its user is that this shouldn't matter if there are many (30) instructions above the VL, in a big region (`HasDistToTop`). And in the case where the scheduled latency is not increased by the VL (another SU at least as high was already scheduled below it) this is also less likely to matter (`PreservesSchedLat`).

I now made the experiment to try your idea, in two versions: The first one checks for 2 free regs, the second one for the pressure to be below half of the limit (it doesn't schedule the SU "low" if the register pressure is ok). If the current pressure was perfectly estimated, at least the second one should give no more spilling. I did get a few more spills/reloads with both versions:

```
patch <> "Pressure + 2 regs < Lim":  +3K spills/reloads
Improvements
0.988: i525.x264_r 
0.991: f538.imagick_r 
0.991: f519.lbm_r 
Regressions
1.035: f544.nab_r 
1.009: i500.perlbench_r 
1.006: f507.cactuBSSN_r 
```

```
patch <> "Pressure <= Lim / 2":  +1K spills/reloads
Improvements
0.990: f519.lbm_r 
Regressions 2017_L_Exp_PSetLimHalf:
1.036: f544.nab_r 
1.007: i505.mcf_r 
1.006: i523.xalancbmk_r 
```

It doesn't seem to give much one way or the other performance-wise... (I also tried another version using the "2 regs rule" only in cases where latency reduction is not enabled, with similar mixed results). Both of these show f544.nab_r regressing. I tried this again on this benchmark, but only when latency reduction is *not* enabled and then the regression dissapeared. There were however only 5 more spill/reloads causing the 3.6% regression, so it's likely not a spilling problem, so not sure exactly what is the cause. 

I agree that this should in theory not hurt if done correctly, and maybe GenericScheduler would do better if it set up the global liveness also. At least it shouldn't have to cause a heavy increase in spilling given that it checks the pressure 'Excess' very early. Another problem in the GenericScheduler relating to this is however that RegPressureTracker:: getUpwardPressureDelta() that is used seems to just track one PressureSet that exceeds its limit. That means e.g. GRX32Bit could be tracked, but then VR16Bit ignored, IIUC. I think it would have to be able to check any Excess that the PressureDiff of an SU would affect. So with global liveness and better handling of pressure Excess it should do better, but then again it will not matter in many interesting cases where there will be Excess anyway where it will always be a matter of reducing it as much as possible. @atrick - have you tried this?

@atrick has described the problem I saw in GenericScheduler as "backing itself into a corner". So the idea for me as been to not be so confident about the overall picture but rather do more as the opportunity arises. First reg-pressure as much as possible, and only after that latency reduction. For bigger regions (like in cactus), there will always be some spilling so in those cases it always makes sense to close the live range, and it seems to give good performance to do it under the given checks.

```
EXPERIMENT:
@@ -256,6 +258,9 @@ int SystemZPreRASchedStrategy::computeSULivenessScore(
     UsesLiveAll = !PrioKill && !GPRKill;
     StoreKill = (PrioKill || (!HasPrioUse && GPRKill));
   } else if (MO0.isReg() && MO0.getReg().isVirtual()) {
+    const RegPressureTracker &RPTracker = DAG->getBotRPTracker();
+    ArrayRef<unsigned> Pressure = RPTracker.getRegSetPressureAtPos();
+    bool InFPReg = false;
     int PrioPressureChange = 0;
     int GPRPressureChange = 0;
     const PressureDiff &PDiff = DAG->getPressureDiff(SU);
@@ -266,12 +271,40 @@ int SystemZPreRASchedStrategy::computeSULivenessScore(
         PrioPressureChange += PC.getUnitInc();
       else if (PC.getPSet() == GPRPressureSet)
         GPRPressureChange += PC.getUnitInc();
+      if (PC.getPSet() == SystemZ::FP16Bit)
+        InFPReg = true;
     }

     if (IsLoad) {
       bool PrioDefNoKill = PrioPressureChange == -RegWeight;
       bool GPRDefNoKill = GPRPressureChange == -RegWeight;
+
+      unsigned PSet = ~0;
+      unsigned RegFactor = 1;
+      if (PrioDefNoKill)
+        PSet = InFPReg ? SystemZ::FP16Bit : SystemZ::VR16Bit;
+      else if (GPRDefNoKill) {
+        PSet = SystemZ::GRX32Bit;
+        RegFactor = 2;
+      }
+      if (PSet != ~0) {
+        unsigned Lim = RPTracker.getRCI()->getRegPressureSetLimit(PSet);
+        if (!RPTracker.getLiveThru().empty())
+          Lim += RPTracker.getLiveThru()[PSet];
+
+        // EXPERIMENTAL:
+        // Guard against some other SU being scheduled instead that could
+        // cause 2 registers to become live. Skip this if it would still be
+        // below the limit.
+        // if (Pressure[PSet] + (2 * RegFactor) < Lim)
+        //   return 0;
+        // ~ or ~
+        // Alternatively, skip if pressure is low.
+        // if (Pressure[PSet] <= Lim / 2)
+        //   return 0;
+      }
+
       UsesLivePrio =

```

JonPsson1 commented:
Some minor fixes. It is all looking quite good now to me so the next step would be to decide on using PressureDiffs (or not). 

Given that the PressureDiffs give identical results, they make sense to me, I guess. I also think that maybe SystemZPreRASchedStrategy::initializePressureSets() could be simplified, given that GCNSchedStrategy::pickNodeFromQueue() is hard-coding these sets by name, which is an indicator that these may be fairly stable.

JonPsson1 commented:
Rebased: The register coalescer has become more active and there are now LAs
defining a PReg (out-arg), which there was not before. biasPhysRegExtra() has been
updated to accept these as well (without the check), and it seems to be no problem.

The hard limit for tiny-region has been replaced by these points:
 - Don't pull down any of 12 highest nodes for the purpose of closing the live range.
 - Demand at least 4 sequenced nodes for latency reduction (two chains or one long chain).
 - Help cmp0 elimination.

This worked well when enabling register pressure tracking for all regions. GenericScheduler
limits this around ~8 (half the number of allocatable GPRs), for the purpose of compile time
reduction. I have not spent that much time on compile time, but it is clear that this limit
is useful:

```
Compile time:         Average Wall   Worst   // Building SPEC with -j1 once per clang version.
GenericScheduler  :      1.82%       23.1%  16.8% 13.9% ...
>= 8 instructions :      1.97%       23.0%  20.0% 13.9% ...
all regions       :      2.88%       23.3%  20.7% 13.9% ...
```

Cmp0 elimination stats over SPEC:
```
systemz-elim-compare - Number of eliminated comparisons
main:                   191062
patch, without CmpElim: 190827
patch, With CmpElim:    192563
```

While this version achieves (all achieved) performance improvements, there are still possible
follow-ups:
 - try setting up global liveness and fine-tuning liveness reduction with RPTracker.
 - Try using DFS subtrees for liveness reduction.
 - Revisit PRegs (bias, Copy2PRegBias)
 - shouldTrackPressure for all single block loops and acyclic-limited latency reduction?

Tests updated and passing.

JonPsson1 commented:
HighSUs has been improved for compile time:
 - don't grow a potentially huge set when only at most 12 SUs should result.
 - use std::sort on a vector instead of inserting all elements one by one.
 - Handle the case of a region with 12 instructions or less by means of having the set empty.

This brings the average pass time down to 1.905%.

Cases with more >10% compile time seemed to be mostly about buildDAGWithRegPressure.

In a smaller (and more common region) perf indicated isSchedLowCand() as begin slow. I tried to cache the result of computeSULivenessScore(), but to my surpise that was slower. I am still experimenting with different versions but I am not sure there is a reasonable fix - it seems it is on par with the GenericSched and maybe it isn't worth cluttering the code too much...

JonPsson1 commented:
Patch updated here after rebase (Dec 4).

Refactoring with NFC against (C) above: I tried to make it more clear how the progression of checks is made in tryCandidate() by refactoring away computeSULivenessScore() and computing the properties for the two candidates in a more clear way.

The logical start was to check if both SUs increase the scheduled latency like this (immediately after biasPhysRegExtra()):

```
  // Both SUs increase the scheduled latency.
  if (!PreservesSchedLat_Cand && !PreservesSchedLat_TryC &&
      shouldReallyReduceLatency()) {
    tryLess(TryCand.SU->getHeight(), Cand.SU->getHeight(), TryCand, Cand,
            GenericSchedulerBase::BotHeightReduce);
    return TryCand.Reason != NoCand;
  }

```
This however doesn't make any difference, so I removed it again: 1) The store in a stores-group doesn't ever increase the scheduled latency (first in group already scheduled w/ same Depth). 2) The "sched low" for liveness is only done if one of the candidates preserves scheduled latency. 

@uweigand Does this look like an improvement?

JonPsson1 committed:
Simplified - no liveness reduction.

JonPsson1 committed:
Simplified - no liveness reduction (NFC to 1449e29).
Tests updated (two using -generic-sched to pass).

JonPsson1 committed:
Revert "Simplified - no liveness reduction."

This reverts commit 1449e29a57c0fd6dad8671c568aaeb00f81246b1.

JonPsson1 committed:
Disable pre-RA MISched.

JonPsson1 committed:
Revert "Disable pre-RA MISched."

This reverts commit 61ddf3f0ef9646c7cf6f0459d964a74371c37952.

uweigand commented on llvm/lib/Target/SystemZ/SystemZMachineScheduler.cpp:
```diff
@@ -21,6 +13,188 @@ using namespace llvm;
+  // data sequences that should be interlaved. These are SUs that only have
+  // one data predecessor / successor edge(s) to their adjacent instruction(s)
+  // in the input order. Disable if region has many SUs relative to the
+  // overall height.
```
I guess I still don't fully understand this heuristics - what is so special about single-sucessor edges?

What if we just do the first check, whether total number of instructions significantly exceeds max height, and just skip the whole data-sequences loop?   Would this still work?

JonPsson1 replied:
I checked the build of SPEC and compared between versions, first comparing main (GenericScheduler) with disabled (no pre-RA sched) and the new strategy. Then also with this current simplification without the data-sequences heuristic. This last step clearly degrades the statistics more than any other version, although only with a few percent - most notably the cmp0 eliminations. As this comes from the more aggressive use of the latency heuristic, I tried moving that down to make it active fewer times, but this did not help. Finally, I then compared with removing the cmp0 handling altogether, and it seems that this did not hurt much as it wasn't helping much with the increased latency scheduling. There doesn't seem to be any performance differences either between the last four versions in the summaries below:

```
* Statistics:

regalloc - Number of spilled live ranges
161454               genericscheduler
160099               disabled
161524               New strategy (simplified)
166553               Same, but without data-sequences heuristic. (+3%)
166552                 - Latency heuristic below weak edges (1 step down).
166600                 - Latency heuristic below cmp/0 elim (2 steps down).
166553                 - Cmp/0 elim removed (only change).

systemz-elim-compare - Number of eliminated comparisons:
191698               GenericScheduler
191843               disabled
193292               New strategy (simplified)
184567               Same, but without data-sequences heuristic. (-5% / -4%)
184567                 - Latency heuristic below weak edges (1 step down).
188133                 - Latency heuristic below cmp/0 elim (2 steps down).
184386                 - Cmp/0 elim removed (only change).

* Instruction mix:

Number of reg move insructions:
888610               GenericScheduler
907157               disabled
888621               New strategy (simplified)
893878               Same, but without data-sequences heuristic. (+1%)
893870                 - Latency heuristic below weak edges (1 step down).
894021                 - Latency heuristic below cmp/0 elim (2 steps down).
893880                 - Cmp/0 elim removed (only change).

Number of spill/restore instructions:
470795               GenericScheduler
467396               disabled
470418               New strategy (simplified)
476687               Same, but without data-sequences heuristic. (+1%)
476686                 - Latency heuristic below weak edges (1 step down).
476802                 - Latency heuristic below cmp/0 elim (2 steps down).
476697                 - Cmp/0 elim removed (only change).

```

uweigand reviewed (commented):
Thanks for the simplified version - this does look like a good step forward, and something I could see approving.  Just one follow-up comment on the data sequence heuristics.

JonPsson1 committed:
Try without HasDataSequences, with tests updates.

JonPsson1 committed:
Revert "Try without HasDataSequences, with tests updates."

This reverts commit 3c427010542ca60ec0291ce92d7956ff0901863c.

uweigand reviewed (approved):
OK, given the performance results I agree we should go with this version for now.  Thanks for exploring all those options!

nikic commented:
https://github.com/llvm/llvm-project/commit/bf2150031a552bd39676336a96da08d8adda27cc should fix the test failures.
